lib/assert.js

```diff
@@ -931,6 +931,88 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+const mustCallChecks = [];
+
+function runCallChecks(exitCode) {
+  if (exitCode !== 0) return;
+
+  const failed = mustCallChecks.filter(function(context) {
+    if ('minimum' in context) {
+      context.messageSegment = `at least ${context.minimum}`;
+      return context.actual < context.minimum;
+    } else {
+      context.messageSegment = `exactly ${context.exact}`;
+      return context.actual !== context.exact;
+    }
+  });
+
+  failed.forEach(function(context) {
+    console.log('Mismatched %s function calls. Expected %s, actual %d.',
+                context.name,
+                context.messageSegment,
+                context.actual);
+    console.log(context.stack.split('\n').slice(2).join('\n'));
+  });
+
+  if (failed.length) process.exit(1);
+}
+
+assert.mustCall = function mustCall(fn, exact) {
+  return _mustCallInner(fn, exact, 'exact');
+};
+
+assert.mustCallAtLeast = function mustCallAtLeast(fn, minimum) {
+  return _mustCallInner(fn, minimum, 'minimum');
+}
+
+function _mustCallInner(fn, criteria = 1, field) {
+  if (process._exiting)
+  // eslint-disable-next-line no-restricted-syntax
+    throw new Error('Cannot use common.mustCall*() in process exit handler');
+  if (typeof fn === 'number') {
+    criteria = fn;
+    fn = noop;
+  } else if (fn === undefined) {
+    fn = noop;
+  }
+
+  if (typeof criteria !== 'number')
+  // eslint-disable-next-line no-restricted-syntax
+    throw new TypeError(`Invalid ${field} value: ${criteria}`);
+
+  // eslint-disable-next-line no-restricted-syntax
+  const stackTrace = util.inspect(new Error());
+
+  const context = {
+    [field]: criteria,
+    actual: 0,
+    stack: stackTrace,
+    name: fn.name || '<anonymous>'
+  };
+
+  // Add the exit listener only once to avoid listener leak warnings
+  if (mustCallChecks.length === 0) process.on('exit', runCallChecks);
+
+  mustCallChecks.push(context);
+  return function() {
+    // Increment cal count
+    context.actual++;
+    if (context.actual === (context.exact || context.minimum)) {
+      // Once function has reached its call count remove it from
+      // MustCallChecks list to prevent memory leaks
+      const index = mustCallChecks.indexOf(context, 0);
+      mustCallChecks.splice(index, 1);
```

Comment:
Later we might consider making these a linked list to make this slightly more efficient

Reply:
… or a `Set`? I think that would also work quite well for now and is more efficient than `.splice()`

Reply:
After switching to a set:

```suggestion
      mustCallChecks.delete(context)
```